drivers/adc/adc_nrfx_saadc.c

```diff
@@ -526,6 +583,13 @@ static int start_read(const struct device *dev,
 	adc_context_start_read(&m_data.ctx, sequence);
 
 	error = adc_context_wait_for_completion(&m_data.ctx);
+	if (error) {
+		return error;
+	}
+
+	if (selected_channels & m_data.single_ended_channels) {
+		correct_single_ended(sequence, active_channels);
```

Comment:
This will not work for `adc_read_async()` because then `adc_context_wait_for_completion()` will most likely return before the samplings are done. Another problem will be with sequences that use a callback - at the time the callback is called, the values will not be corrected yet.
I think the correction should be done in `saadc_irq_handler()` instead.

Reply:
Right :) Moved correction to irq handler.